File: utilities/src/main/java/xerus/util/helpers/Rater.java
package xerus.util.helpers;

public class Rater<X> {
	
	public X obj;
	public double points;
	protected boolean inv;
	
	public Rater() {
		this(false);
	}
	
	public Rater(boolean invert) {
		this(null, invert ? Double.MAX_VALUE : Double.MIN_VALUE, invert);
	}
	
	public Rater(X object, double p) {
		this(object, p, false);
	}
	
	public Rater(X object, double p, boolean invert) {
		this.obj = object;
		points = p;
		inv = invert;
	}
	
	public boolean update(Rater<X> other) {
		return update(other.obj, other.points);
	}
	
	public boolean update(Rater<X> other, double multiplier) {
		return update(other.obj, other.points * multiplier);
	}
	
	/** replaces the objects if the given points are higher than the saved ones */
	public boolean update(X newobj, double newpoints) {
		if ((!inv && newpoints > points) || (inv && newpoints < points)) {
			obj = newobj;
			points = newpoints;
			return true;
		}
		return false;
	}
	
	public boolean update(X newobj, double newpoints, boolean bonus) {
		if (bonus) newpoints++;
		return update(newobj, newpoints);
	}
	
	public boolean hasobj() {
		return obj != null;
	}
	
	public String toString() {
		return obj.toString() + " - Punkte: " + points;
	}
	
	public boolean equals(Object obj) {
		if (obj.getClass() != this.getClass())
			return false;
		Rater m = (Rater) obj;
		return m.points == this.points && m.obj.equals(obj);
	}
	
	public int hashCode() {
		return obj.hashCode() * 9 + (int) points;
	}
	
}

File: kotlin/src/xerus/ktutil/helpers/Rater.kt
package xerus.ktutil.helpers

open class Rater<X> constructor(
		/** the current object of this Rater */
		@JvmField
		var obj: X?,
		/** the points to the current [obj] of this Rater */
		@JvmField
		var points: Double,
		/** if true, then objects with less points will be preferred */
		private val inverted: Boolean = false) : Comparable<Rater<X>> {
	
	constructor(invert: Boolean = false) : this(null, originalPoints(invert), invert)
	
	fun hasObj() = obj != null
	
	fun clear() {
		points = originalPoints(inverted)
		obj = null
	}
	
	/** replaces the object if the given points are higher than the saved ones
	 * @return if [other] became the current [obj] */
	open fun update(other: X?, otherPoints: Double): Boolean {
		if (!inverted && otherPoints > points || inverted && otherPoints < points) {
			obj = other
			points = otherPoints
			return true
		}
		return false
	}
	
	fun update(other: Rater<X>, multiplier: Double = 1.0) = update(other.obj, other.points * multiplier)
	
	override fun compareTo(other: Rater<X>) = points.compareTo(other.points)
	
	override fun toString() = "%s - Points: %.2f".format(obj, points)
	
	override fun equals(other: Any?) =
			other is Rater<*> && other.points == this.points && other.obj == obj
	
	override fun hashCode() = obj?.hashCode() ?: 0 * 9+points.toInt()
	
	companion object {
		private fun originalPoints(inverted: Boolean) = if (inverted) Double.POSITIVE_INFINITY else Double.NEGATIVE_INFINITY
	}
	
}
